File: framework/languages/dotnet/e2e/RegistryMessaging/Client/Scenarios/RmC8PayloadRoundTripScenario.cs
using System.Security.Cryptography;
using System.Text;
using RegistryMessaging.Client.Support;
using RegistryMessaging.Shared;
using Zlink.HttpClient;

namespace RegistryMessaging.Client.Scenarios;

// RM-C8 verifies payload size round trips by checking returned length and hash
// for several message sizes.
internal static class RmC8PayloadRoundTripScenario
{
    public static async Task RunAsync(ZLinkHttpClient singleConsumer, ZLinkHttpClient providerA)
    {
        var before = providerA.Get("/evidence").Fetch<string[]>();
        var markers = new List<string>();
        foreach (var size in new[] { 1, 4096, 256 * 1024, 1024 * 1024 })
        {
            var marker = $"rm-c8-{size}-{Guid.NewGuid():N}";
            markers.Add(marker);
            var payload = BuildPayload(size);
            var expectedHash = HashPayload(payload);
            var reply = (await singleConsumer.Post("/profile/payload")
                .Body(new PayloadReq(marker, payload))
                .SubmitAsync<PayloadRes>()).Body;
            ScenarioAssert.That(reply.Marker == marker, "RM-C8 marker mismatch.");
            ScenarioAssert.That(reply.Length == payload.Length, "RM-C8 payload length mismatch.");
            ScenarioAssert.That(reply.Sha256 == expectedHash, "RM-C8 payload hash mismatch.");
        }

        var followUp = (await singleConsumer.Post("/profile/request")
            .Body(new ProfileReq("rm-c8-after"))
            .SubmitAsync<ProfileRes>()).Body;
        ScenarioAssert.That(followUp.Value == "profile:rm-c8-after", "RM-C8 follow-up request failed.");

        var after = providerA.Get("/evidence").Fetch<string[]>();
        ScenarioAssert.That(
            markers.All(marker =>
                ScenarioAssert.CountNewEvidence(after, before, "payload-request|rid=api-a", marker) == 1),
            "RM-C8 payload evidence missing.");
        Console.WriteLine("scenario RM-C8 passed");
    }

    private static string BuildPayload(int size)
    {
        var builder = new StringBuilder(size);
        for (var i = 0; i < size; i++) builder.Append((char)('a' + i % 26));

        return builder.ToString();
    }

    private static string HashPayload(string payload)
    {
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(payload)));
    }
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/client/Scenarios/RmC8PayloadRoundTripScenario.kt
package systems.zlink.e2e.kotlin.registrymessaging.client.Scenarios

import java.security.MessageDigest
import java.util.HexFormat
import java.util.UUID
import systems.zlink.e2e.kotlin.registrymessaging.client.Support.HttpJson
import systems.zlink.e2e.kotlin.registrymessaging.client.Support.ScenarioAssert
import systems.zlink.e2e.kotlin.registrymessaging.shared.PayloadRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.PayloadReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.RequestFailureRes

object RmC8PayloadRoundTripScenario {
    fun run(singleConsumer: HttpJson, providerA: HttpJson, providerB: HttpJson) {
        val markers = mutableListOf<String>()
        for (size in listOf(1, 4096, 256 * 1024, 1024 * 1024)) {
            val marker = "rm-c8-$size-${UUID.randomUUID().toString().replace("-", "")}"
            markers += marker
            val payload = buildPayload(size)
            val reply = singleConsumer.post<PayloadRes>("/profile/payload", PayloadReq(marker, payload))
            ScenarioAssert.that(reply.marker == marker, "RM-C8 marker mismatch.")
            ScenarioAssert.that(reply.length == payload.length, "RM-C8 payload length mismatch.")
            ScenarioAssert.that(reply.sha256 == hash(payload), "RM-C8 payload hash mismatch.")
        }
        val oversized = singleConsumer.post<RequestFailureRes>(
            "/profile/payload-over-limit",
            PayloadReq("rm-c8-over-limit-${UUID.randomUUID()}", buildPayload(3 * 1024 * 1024)),
        )
        ScenarioAssert.that(oversized.failed, "RM-C8 oversized payload should fail.")
        ScenarioAssert.that(
            oversized.failureType == "TimeoutException",
            "RM-C8 oversized payload should report TimeoutException, got ${oversized.failureType}.",
        )
        val followUp = singleConsumer.post<ProfileRes>("/profile/request", ProfileReq("rm-c8-after"))
        ScenarioAssert.that(followUp.value == "profile:rm-c8-after", "RM-C8 follow-up request failed.")
        val evidence = providerA.get<List<String>>("/evidence") + providerB.get<List<String>>("/evidence")
        ScenarioAssert.that(
            markers.all { marker -> evidence.any { it.contains("payload-request") && it.contains(marker) } },
            "RM-C8 payload evidence missing.",
        )
        println("scenario RM-C8 max-message-size passed")
    }

    private fun buildPayload(size: Int): String =
        buildString(size) {
            for (index in 0 until size) {
                append(('a'.code + (index % 26)).toChar())
            }
        }

    private fun hash(payload: String): String =
        HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(payload.toByteArray(Charsets.UTF_8))).uppercase()
}
